autodiff/reverse/var/var.hpp

```diff
@@ -911,19 +1058,168 @@ struct Hypot3Expr : TernaryExpr<T>
 
     Hypot3Expr(const T& v, const ExprPtr<T>& ll, const ExprPtr<T>& cc, const ExprPtr<T>& rr) : TernaryExpr<T>(v, ll, cc, rr) {}
 
-    virtual void propagate(const T& wprime)
+    void propagate(const T& wprime) override
     {
         l->propagate(wprime * l->val / val);
         c->propagate(wprime * c->val / val);
         r->propagate(wprime * r->val / val);
     }
 
-    virtual void propagatex(const ExprPtr<T>& wprime)
+    void propagatex(const ExprPtr<T>& wprime) override
     {
         l->propagatex(wprime * l / val);
         c->propagatex(wprime * c / val);
         r->propagatex(wprime * r / val);
     }
+
+    void update() override
+    {
+        l->update();
+        c->update();
+        r->update();
+        this->val = hypot(l->val, c->val, r->val);
+    }
+};
+
+// Any expression yielding a boolean depending on arithmetic subexpressions
+struct BooleanExpr
+{
+    template<typename Comparator>
+    struct Negate {
+      template<typename T>
+      bool operator () (const T& a, const T& b) const {
+        return !(Comparator {}(a, b));
+      }
+    };
+
+    bool val = {};
+
+    explicit BooleanExpr(bool v) : val(v) {}
+    operator bool() const { return val; }
+
+    virtual void update() = 0;
+    virtual std::shared_ptr<BooleanExpr> wrap() const = 0;
+};
+
+/// Capture numeric comparison between two expression trees
+template<typename T, typename Comparator>
+struct ExpressionComparison : public BooleanExpr
+{
+    ExprPtr<T> l;
+    ExprPtr<T> r;
+    Comparator comparator;
+
+    explicit ExpressionComparison(const ExprPtr<T>& ll, const ExprPtr<T>& rr, Comparator&& compare)
+        : BooleanExpr(compare(ll->val, rr->val)), l(ll), r(rr), comparator(compare) {}
+
+    void update() override
+    {
+        l->update();
+        r->update();
+        val = comparator(l->val, r->val);
+    }
+
+    std::shared_ptr<BooleanExpr> wrap() const override { return std::make_shared<ExpressionComparison>(*this); }
+
+    auto operator! () const { return ExpressionComparison<T, Negate<Comparator>>(l, r, Negate<Comparator> {}); }
+};
+
+// CTAD for ExpressionComparison
+template<typename T, typename Comparator>
+ExpressionComparison(const ExprPtr<T>&, const ExprPtr<T>&, Comparator&&) -> ExpressionComparison<T, Comparator>;
+
+/// Capture boolean operations between boolean expressions
+template<typename Operation>
+struct BooleanOperation : public BooleanExpr
```

Comment:
Can't we define operators `||`, `&&`, etc. in a way that they produce a new `BooleanExpr`, without creating this additional type, `BooleanOperation`? 

Just trying to see if there is an alternative approach with less indirection. If not possible, then OK.

Reply:
It's possible that I'm overlooking some corner cases here, in which these additional types would provide a more general solution (e.g., mixing `BooleanExpr` with `bool`).

Reply:
In principle I could see all of those use cases being encompassed by lambdas. I can try it out if you like, but it would mean a different paradigm for this part of the implementation when compared to the existing bits

Reply:
I'm not totally sure that lambda functions would eliminate the need for these additional types, but they are powerful constructs not to be underestimated. If this becomes too much work, however, let's go with your current approach.